src/tlo/population.py

```diff
@@ -1,39 +1,67 @@
 """The Person and Population classes."""
 
 import math
-from typing import Any, Dict
+from types import TracebackType
+from typing import Any, Dict, Optional, Set, Type
 
 import pandas as pd
 
-from tlo import logging
+from tlo import Property, logging
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
 
-class PatientDetails:
-    """Read-only memoized view of population dataframe row."""
-    
-    def __init__(self, population_dataframe: pd.DataFrame, person_id: int):
+class IndividualProperties:
+    """Memoized view of population dataframe row that is optionally read-only."""
+
+    def __init__(
+        self,
+        population_dataframe: pd.DataFrame,
```

Comment:
Interesting question as to whether we should hold reference to population instance or props here. Insisting on using this class with a context manager should discourage holding the object for long periods of time. That'll be a problem because we switch out the population's props dataframe when we add rows for growing population.

Reply:
I've now refactored the code a bit to simplify allowing 'locking' the population dataframe when withing the context manager in future without having to store a reference to population in individual properties. The `Population.individual_properties` method itself now acts as a context manager (rather than the returned object) and so can easily in future do something like set a `self._locked` attribute which if true prevents direct access to the dataframe via the `props` attribute (making it a property) and also disallows `do_birth` being called. The object returned by the `Population.individual_properties` context manager will now also be 'finalized' on exiting the context, so that any subsequent attempts to read or update properties raise an error. The object also no longer stores even a 'private' reference to the population dataframe (with access instead mediated via closures defined in initialiser) which should further guard against direct accesses to population dataframe.